### PR TITLE
Fix copying simple_rte_array in WITHIN GROUP planning.

### PR DESCRIPTION
**Pull request by hlinnaka**
When making a temporary copy of PlannerInfo, must make a deep copy of
'simple_rte_array', just like we did for 'simple_rel_array'.

Test included, which caused a crash/assertion failure before this fix.

### CODE REVIEW

**Comment by hlinnaka** [comment]
5X_STABLE only. This bug doesn't exist in 'master', because the WITHIN GROUP planning code has been completely rewritten, to use the upstream code.

File: src/test/regress/expected/bfv_olap.out [hunk 1]
```diff
@@ -522,6 +522,129 @@ WHERE c = 87 ;
 ----------
 (0 rows)
 
+--
+-- This used to crash or fail an assertion, because 'simple_rte_array' was not
+-- copied correctly in WITHIN GROUP planning, causing reference-after-free.
+--
+CREATE TABLE hundred_parts(a int, b int, c int, d int) DISTRIBUTED RANDOMLY
+  PARTITION BY RANGE (a) (START (1) END (100) EVERY (1));
+INSERT INTO hundred_parts SELECT g, g/10, g/20, g/25 from generate_series(1,99) g;
```
Comment (BaiShaoqi) [inline src/test/regress/expected/bfv_olap.out]: Learned how to insert multiple columes to a table by `generate_series`.

First time to see a sample of `WITHIN GROUP`. Never know this before.

File: src/test/regress/expected/bfv_olap.out [hunk 2]
```diff
@@ -522,6 +522,129 @@ WHERE c = 87 ;
...
+       percentile_disc(0.90) WITHIN GROUP (ORDER BY c) AS perc90,
+       percentile_disc(0.80) WITHIN GROUP (ORDER BY d) AS perc80,
+       percentile_disc(0.70) WITHIN GROUP (ORDER BY b) AS perc70,
+       percentile_disc(0.60) WITHIN GROUP (ORDER BY c) AS perc60,
+       percentile_disc(0.50) WITHIN GROUP (ORDER BY d) AS perc50,
+       percentile_disc(0.40) WITHIN GROUP (ORDER BY b) AS perc40,
+       percentile_disc(0.30) WITHIN GROUP (ORDER BY c) AS perc30,
+       percentile_disc(0.20) WITHIN GROUP (ORDER BY d) AS perc20,
+       percentile_disc(0.10) WITHIN GROUP (ORDER BY b) AS perc10
+FROM hundred_parts
```
Comment (BaiShaoqi) [inline src/test/regress/expected/bfv_olap.out]: Do you have a special meaning to select so many columns ?
Will the following two columns not work ?
```
SELECT a,
       percentile_disc(1) WITHIN GROUP (ORDER BY b) AS perc100
GROUP BY a;
```

Comment (BaiShaoqi) [reply src/test/regress/expected/bfv_olap.out]: Yes. Thanks

File: src/test/regress/expected/bfv_olap.out [hunk 3]
```diff
@@ -522,6 +522,129 @@ WHERE c = 87 ;
 ----------
 (0 rows)
 
+--
+-- This used to crash or fail an assertion, because 'simple_rte_array' was not
+-- copied correctly in WITHIN GROUP planning, causing reference-after-free.
+--
+CREATE TABLE hundred_parts(a int, b int, c int, d int) DISTRIBUTED RANDOMLY
+  PARTITION BY RANGE (a) (START (1) END (100) EVERY (1));
```
Comment (melanieplageman) [inline src/test/regress/expected/bfv_olap.out]: Is the difference between this and the existing tests for `WITHIN GROUP` in src/test/regress the partition table? If so, does this issue only manifest for a partition table?

Comment (hlinnaka) [reply src/test/regress/expected/bfv_olap.out]: Yeah, all the existing tests seem to have a single non-partitioned table in the FROM clause.

The important detail here is that the range table contains a lot of entries. Partitioning is one way to accomplish that, but I think you would get the same effect with a lot of joins, for example.

**Comment by d** [comment]
I took a glance at this fix, my interpretation is: planner will crash while planning `WITHIN GROUP` on an inheritance hierarchy. But I still don't get when this would happen only to `RANDOMLY DISTRIBUTED` tables.

**Comment by hlinnaka** [comment]
> I took a glance at this fix, my interpretation is: planner will crash while planning `WITHIN GROUP` on an inheritance hierarchy. But I still don't get when this would happen only to `RANDOMLY DISTRIBUTED` tables.

The crash happens when the planner creates a two-stage Agg plan. If the table is distributed by the GROUP By column, then the planner can create a single-stage Agg plan instead, which doesn't exercise the buggy code. Or maybe it's creating a three-stage Agg plan vs two-stage, not sure. But you get a different plan, in any case.

**Comment by hlinnaka** [comment]
Pushed, thanks for the reviews!

**Pull request merged**